### ISSUE TITLE
Notification before expiry date

### ISSUE THREAD

**Issue by GlitchCat** [TheLastProject/Catima#410]
Catima allows setting an expiry date. It would be useful to get a notification before the card is expired, in the case of cards with monetary value.

Could be implemented by setting an per-card notification or by adding it to the settings. 
Like being able to set a notification for x weeks or days before the expiry date.

**Comment by Ashutoshgupta22** [comment]
Hey !  I would like to work on this issue ..

**Comment by Ashutoshgupta22** [comment]
Hey!  I would like to work on this issue ..

**Comment by TheLastProject** [comment]
Sure, go ahead. I can't help you much with this though, as Android "alarms" are not existing knowledge to me either.

**Comment by TheLastProject** [comment]
Someone on Google Play asked to add it to a calendar. And that actually raised an interesting idea: instead of implementing this as alerts and everything in Catima, Catima would become a Calendar Provider: https://developer.android.com/guide/topics/providers/calendar-provider

It could then automatically expose a Catima calendar with all expiry dates in there and if cards get deleted/updated update it in the calendar.

Not sure if easier, may be way harder, but it'd be interesting to look into.

**Comment by TheLastProject** [comment]
Of course, a much simpler method would be to, when you save a card with a change in expiry date, ask the user if they want to add that expiry date to their calendar.

**Comment by TheLastProject** [comment]
Actually, let's go for that, it would be pretty simple to implement and very useful. Users don't expect things shared to their calendar to be synced up normally anyway.

Changing this task to "Good first issue" because implementing it that way would not be much more than:
1. See if the expiry date changed when someone hits save
2. Show a dialog
3. If yes, start a calendar intent

**Comment by Chaynes1** [comment]
I see the pull request was never merged? I would like to work this issue if that is cool.

**Comment by TheLastProject** [comment]
Yeah that PR was never finished.

Looking back, I'm not 100% sure which method of implementing would be best, but you're free to give implementing it a try :)

**Comment by Chaynes1** [comment]
So I can do something similar to this


![](https://user-images.githubusercontent.com/68068357/137925824-703464f2-82cd-4b58-9fd8-78bea2b6402c.png)

And just have the app sync those reminders with the Calendar using the Calendar Provider API you mentioned above?

**Comment by TheLastProject** [comment]
There are many ways to go about it and in typical me fashion I've probably been overcomplicating it.

I think we need to go back to the design phase and use something that's simpler and less flexible, but should still work perfectly in the vast majority of use cases.

Maybe a better way to design this would be to:
1. Add a section to the settings screen for reminders. In this section, add:
    - An "Enable reminders" checkbox. If disabled, all other options are greyed out.
    - A "Remind me for expiring cards" which is just a bunch of checkboxes like "Day of expiry", "1 day in advance", "2 days in advance", "3 days in advance", "1 week in advance", "1 month in advance" (wonder which are the best possible values, maybe these are? We can easily add more later)
    - A "Reminder time" which allows you to select at what hour you want reminders (what is a good default? 11:00 maybe, as most people probably are awake by then but it's still early enough to get a reminder for a lunch coupon being about to expire?)
2. When enabling reminders or changing the reminder time set an [Android timer](https://developer.android.com/reference/java/util/Timer) (assuming this is the right way) for the chosen time and cancel all other Catima-set timers.
3. When the timer is called, loop over all cards, check the expiry field and compare it to the expiry alert setting.

I have no experience with Android timers though, so this may be quite a challenging task (how do we make sure Android will actually execute our timer?), but also probably closest to how people expect apps to behave (the app itself reminds you)

Do you have any suggestions or opinions?

**Comment by Chaynes1** [comment]
Yea, Ok that is fair. I didn't know how much you wanted to deviate from what you and the previous guy were working....

Yea, I like that idea. Seems simple enough. For "Remind me for expiring cards" I don't know which would be better either. (If having multiple or one reoccurring one) The core logic should be done where this would be just a detail and change when we see it...

Yea, I don't have much experience with Android notifications, but my only suggestion would be to follow the standard way with how most apps send notifications, instead of updating the users calendar. I have one app that does that and it seems kinda invasive. (At least to me)

Just briefly searching I saw mention of an AlarmManager to schedule future intents to wake your app up. We could have some broadcastReceiver listening for this intent to push a notification to the user.

**Comment by TheLastProject** [comment]
> Yea, Ok that is fair. I didn't know how much you wanted to deviate from what you and the previous guy were working....

I'm always willing to change plans if the new plan is better, especially if it hasn't been released yet as there won't be anyone used to the way it works yet. Designing stuff is hard, and sometimes even when the plan seems perfect it turns out the actual experience isn't that great. But I'm pretty hopeful about this one :)

> Yea, I like that idea. Seems simple enough. For "Remind me for expiring cards" I don't know which would be better either. (If having multiple or one reoccurring one) The core logic should be done where this would be just a detail and change when we see it...

Makes sense to me, I'd think the cleanest way to do this would be to add a [MultiSelectListPreference](https://developer.android.com/reference/kotlin/androidx/preference/MultiSelectListPreference?hl=en) in `xml/preferences.xml` and have the keys just be things like 1d, 2d, 1w, 2w, 1m, etc. I would expect Android to have something built-in for comparing if something is like 1 day different or 1 week different or whatever, so the letters that should be used should become clear during development and we can just ask Android "Hey, is this 1d difference?".

Personally, I think these options would make sense:

| Option | Selected by default |
|-|-|
| Today | Y |
| In 1 day | Y |
| In 2 days | N |
| In 3 days | N |
| In 4 days | N |
| In 5 days | N |
| In 6 days | N |
| In 1 week | Y |
| In 2 weeks | N |
| In 3 weeks | N |
| In 1 month | N |

But we can always tweak this during development.

> Yea, I don't have much experience with Android notifications, but my only suggestion would be to follow the standard way with how most apps send notifications, instead of updating the users calendar. I have one app that does that and it seems kinda invasive. (At least to me)

That sounds like the best option, I definitely don't want Catima to feel invasive, especially given Catima is designed to be privacy-respecting.

> Just briefly searching I saw mention of an AlarmManager to schedule future intents to wake your app up. We could have some broadcastReceiver listening for this intent to push a notification to the user.

Sounds like the way to go :)
